pkg/injector/pod_patch.go

```diff
@@ -847,6 +848,17 @@ func getSidecarContainer(annotations map[string]string, id, daprSidecarImage, im
 	if resources != nil {
 		c.Resources = *resources
 	}
+
+	if runtime.GOOS == "windows" {
```

Comment:
Rather than checking the OS of the control plane, can you check the OS of the container that is being deployed?

Imagine the case where Dapr's control plane is running on Linux, but it's deploying a Windows container.

Reply:
Actually I was exploring that. So the pod spec has a field called `PodOS`, [documentation](https://pkg.go.dev/k8s.io/api/core/v1#PodSpec), but it requires a special feature flag to be set - `IdentifyPodOS`.

From [Feature Gates | Kubernetes](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
> `IdentifyPodOS`: Allows the Pod OS field to be specified. This helps in identifying the OS of the pod authoritatively during the API server admission time. In Kubernetes 1.24, the allowed values for the `pod.spec.os.name` are `windows` and `linux`.


|Feature| Default|Stage|Since|Until|
-- | -- | -- | -- | --
|IdentifyPodOS | false | Alpha | 1.23 | 1.23
IdentifyPodOS | true | Beta | 1.24 |  

Since Dapr supports versions of k8s < 1.24, we will need to explicitly enable this feature to work. So we can conditionally support this feature based on the feature flag for Windows. What do you think?

Reply:
Uhm that doesn't sound good either.

How about just running the `.cmd` file on every Dapr container? That is, we modify the launch command for *all* Windows containers to make that run before any script. I see that you've already added a line that checks if there's an env var, otherwise it just exits.

I believe this could be done by making the `.cmd` file the ENTRYPOINT of the container. You can still run any command you want which will be passed as shell arg to the `.cmd` file. The `.cmd` file will do what it needs to do, then read the shell arg and invoke that. See example here: https://stackoverflow.com/a/56587335/192024

Reply:
This hypothesis is not turning out to be true. If I have an ENTRYPOINT and I invoke `docker run --rm -it <image name>` it seems to be working, but in the context of k8s, command seems to be overriding the ENTRYPOINT.

From [Define a Command and Arguments for a Container | Kubernetes](https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/),
> The command and arguments that you define in the configuration file override the default command and arguments provided by the container image. If you define args, but do not define a command, the default command is used with your new arguments.
Note: The command field corresponds to entrypoint in some container runtimes. 

Exploring more on this.

Reply:
Uhm I see. What about just setting the command as "args"? See: https://stackoverflow.com/a/67113557/192024

Downsides is that this _may_ require updating the Linux image too (not 100% sure about that)

Reply:
I think for windows case, should better add a doc to guide users on what should do, as considering and handling all situations is hard.

Reply:
Yes, it is easy to implement.

Reply:
> Uhm I see. What about just setting the command as "args"? See: https://stackoverflow.com/a/67113557/192024
> 
> Downsides is that this _may_ require updating the Linux image too (not 100% sure about that)

That does indeed work for Windows. We will need to add an `ENTRYPOINT` for Linux with this approach. Will confirm and update this thread.

Reply:
Let me know what you think of the current approach @ItalyPaleAle @daixiang0 

Going the feature flag route we can skip the entrypoint modification and changing the pod_patch, which might be better since our new entry point will be a shell that handles dapr (as compared to running dapr directly). If we use feature flag, we can set our script to run via the k8s PostStart hook.

Reply:
@shubham1172 excellent work testing this, thanks!! If you confirm that everything works (even on Linux, where our containers do not have `/bin/sh`), then it’s good with me!

Reply:
Thanks @ItalyPaleAle. I confirm that this works on Linux distroless images (without a shell). Also from the docs:
![image](https://user-images.githubusercontent.com/22132944/173399113-c8fe8443-a554-4bb1-a997-cb36f27e8d1f.png)